### PR TITLE
[mesh-diag] new module and new API to discover network topology

### PR DESCRIPTION
**Pull request by abtink**
This commit introduces new module `MeshDiag` which provides new API to discover current network topology.

Information about all routers in mesh are discovered including Router ID, RLOC16, MAC Extended Address, along with the list of established links between routers and their link quality. The API also provides information such as which router is leader, which one is acting as border router and which router is this device itself or parent of this device (if we query from a child). The API allows user to request additional information per router: list of IPv6 addresses per router, or information about children including child RLOC16, mode, and link quality.

The implementation uses multiple Network Diagnostic messages to query routers and then compiles the related information which makes it compatible with Thread 1.1 devices.

Config `OPENTHREAD_CONFIG_MESH_DIAG_ENABLE` and/or `OT_MESH_DIAG` CMake option can be used to enable this feature. It is by default enabled on devices that are configured to act as a border router `ROUTING_MANAGER` is enabled.

This commit also adds a CLI command `meshdiag topology` for this API along with documentation and examples.

---

Example of CLI output (with `ip6-addrs` and `children`):
```
> meshdiag topology children ip6-addrs
id:62 rloc16:0xf800 ext-addr:ce349873897233a5 - me - br
   3-links:{ 46 }
   ip6-addrs:
       fdde:ad00:beef:0:0:ff:fe00:f800
       fdde:ad00:beef:0:211d:39e9:6b2e:4ad1
       fe80:0:0:0:cc34:9873:8972:33a5
   children: none
id:02 rloc16:0x0800 ext-addr:8aa57d2c603fe16c - leader - br
   3-links:{ 46 51 }
   ip6-addrs:
       fdde:ad00:beef:0:0:ff:fe00:fc00
       fdde:ad00:beef:0:0:ff:fe00:800
       fdde:ad00:beef:0:8a36:a3eb:47ae:a9b0
       fe80:0:0:0:88a5:7d2c:603f:e16c
   children:
       rloc16:0x0803 lq:3, mode:rn
       rloc16:0x0804 lq:3, mode:rdn
id:33 rloc16:0x8400 ext-addr:d2e511a146b9e54d
   3-links:{ 57 }
   ip6-addrs:
       fdde:ad00:beef:0:0:ff:fe00:8400
       fdde:ad00:beef:0:824:a126:cf19:a9f4
       fe80:0:0:0:d0e5:11a1:46b9:e54d
   children: none
id:51 rloc16:0xcc00 ext-addr:9aab43ababf05352
   3-links:{ 02 46 57 }
   ip6-addrs:
       fdde:ad00:beef:0:0:ff:fe00:cc00
       fdde:ad00:beef:0:2986:bba3:12d0:1dd2
       fe80:0:0:0:98ab:43ab:abf0:5352
   children: none
id:57 rloc16:0xe400 ext-addr:dae9c4c0e9da55ff
   3-links:{ 33 51 }
   ip6-addrs:
       fdde:ad00:beef:0:0:ff:fe00:e400
       fdde:ad00:beef:0:87d0:550:bc18:9920
       fe80:0:0:0:d8e9:c4c0:e9da:55ff
   children:
       rloc16:0xe402 lq:3, mode:rn - br
       rloc16:0xe403 lq:3, mode:rn
id:46 rloc16:0xb800 ext-addr:fe109d277e0175cc
   3-links:{ 02 51 62 }
   ip6-addrs:
       fdde:ad00:beef:0:0:ff:fe00:b800
       fdde:ad00:beef:0:df4d:2994:d85c:c337
       fe80:0:0:0:fc10:9d27:7e01:75cc
   children: none
Done
```

### CODE REVIEW

File: src/core/utils/mesh_diag.hpp [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+namespace Utils {
+
+/**
+ * This class implements the Mesh Diagnostics.
+ *
+ */
+class MeshDiag : public InstanceLocator
+{
+public:
+    typedef otMeshDiagDiscoverConfig   DiscoverConfig;   ///< The discovery configuration
```
Comment (jwhui) [inline src/core/utils/mesh_diag.hpp:69]:
```suggestion
    typedef otMeshDiagDiscoverConfig   DiscoverConfig;   ///< The discovery configuration.
```

File: src/core/utils/mesh_diag.cpp [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+
+    mDiscoverCallback.InvokeIfSet(error, &routerInfo);
+
+exit:
+    return;
+}
+
+void MeshDiag::HandleTimer(void)
+{
+    // Timed out waiting for response from one or more routers.
```
Comment (jwhui) [inline src/core/utils/mesh_diag.cpp:176]: Should we rely on the CoAP timeout instead? We can configure the CoAP parameters to something other than the default if needed.

Comment (abtink) [reply src/core/utils/mesh_diag.cpp:182]: Yes. We can use the CoAP response timeout and I was about to change this but thinking more on this, I expect we will need a timer for other features/commands we want to add to `MeshDiag`, so may be easier to use the `mTimer` for this as well.

Comment (jwhui) [reply src/core/utils/mesh_diag.cpp:182]: Relying on the CoAP timeout helps ensure that the CoAP transaction has ended.

Alternatively, we could use a separate timeout but should also explicitly cancel the CoAP transaction in the timeout handler.

Comment (abtink) [reply src/core/utils/mesh_diag.cpp:182]: Good point on canceling them. 
I added `Get<Tmf::Agent>().AbortTransacation()` from `Cancel()` but not from timer handler. Will add it.

Comment (jwhui) [reply src/core/utils/mesh_diag.cpp:182]: Given a 5 second timeout and the default CoAP parameters (2 second initial retransmission timeout), there will only be one CoAP retransmission.

Comment (abtink) [reply src/core/utils/mesh_diag.cpp:182]: Yes. I expect so.  Sounds ok to me to allow one retx before overall timeout? What do you think?

- Default coap parm is `ack-timeout = 2 s` with `factor = 3/2 = 1.5`
- So initial wait is randomly selected in `[2, 3]` seconds.
- After than we double the retx/wait time 
  - So even if `2` is picked, first retx after `2`, second one would be at `2+2*2=6` which won't happen if we use overall timeout of `5` sec..

We can define different CoAP params for this, but not sure if we can have better values:
- One thing that I sort of not like about CoAP retx model, is the initial random interval that is picked will be reused for all next attempts (same value is doubled). 
- So a device picking shorter initial time will retx faster and finish earlier (unless we use `factor` very close to `1` which does not sound good).

Comment (jwhui) [reply src/core/utils/mesh_diag.cpp:182]: I think this is fine for now. We can always adjust later as we get more operational experience.

File: include/openthread/mesh_diag.h [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+
+/**
+ * This structure represents the set of configurations used when discovering mesh topology indicating which items to
+ * discover.
+ *
+ */
+typedef struct otMeshDiagDiscoverConfig
+{
+    bool mDiscoverIp6Addresses : 1; ///< Whether or not to discover IPv6 addresses of every router.
+    bool mDiscoverChildTbale : 1;   ///< Whether or not to discover children of every router.
```
[the remark this answers is not in the document]
Comment (abtink) [reply include/openthread/mesh_diag.h]: Yes. Indeed. Thanks.
(sublime editor smart auto-complete ensures the typo persist 😄)

File: src/cli/cli.cpp [hunk 4]
```diff
@@ -5411,6 +5411,168 @@ template <> otError Interpreter::Process<Cmd("nexthop")>(Arg aArgs[])
...
     return error;
 }
+
+#if OPENTHREAD_CONFIG_MESH_DIAG_ENABLE
+
+template <> otError Interpreter::Process<Cmd("meshdiag")>(Arg aArgs[])
+{
+    otError error = OT_ERROR_NONE;
+
+    if (aArgs[0] == "topology")
```
Comment (Vyrastas) [inline src/cli/cli.cpp:5421]: We should add Doxygen tags for the CLI reference by default whenever a new command is added.  My team can do this later, though, if this needs to get in sooner.

Comment (abtink) [reply src/cli/cli.cpp:5472]: Added copy of docs from `README.md` here. It contains bullet point list with sub-lists - not sure if sub-lists will work and look ok. If not we can fix later.

File: src/cli/cli.cpp [hunk 5]
```diff
@@ -5416,6 +5416,214 @@ template <> otError Interpreter::Process<Cmd("nexthop")>(Arg aArgs[])
...
+     *    3-links:{ 33 57 }
+     *    2-links:{ 46 }
+     * id:57 rloc16:0xe400 ext-addr:dae9c4c0e9da55ff
+     *    3-links:{ 46 51 }
+     *    1-links:{ 33 }
+     * Done
+     * @endcode
+     * @par
+     * Discover network topology (list of routers and their connections).
+     *
```
Comment (Vyrastas) [inline src/cli/cli.cpp:5446]: We can't have blank lines in the cli blocks.  Can you put `@par` here instead

File: src/cli/cli.cpp [hunk 6]
```diff
@@ -5416,6 +5416,214 @@ template <> otError Interpreter::Process<Cmd("nexthop")>(Arg aArgs[])
...
+     * Parameters are optional and indicate additional items to discover. Can be added in any order.
+     * * `ip6-addrs` to discover the list of IPv6 addresses of every router.
+     * * `children` to discover the child table of every router.
+     *
```
Comment (Vyrastas) [inline src/cli/cli.cpp:5450]: Should be `@par`

File: src/cli/cli.cpp [hunk 7]
```diff
@@ -5416,6 +5416,214 @@ template <> otError Interpreter::Process<Cmd("nexthop")>(Arg aArgs[])
...
 }
+
+#if OPENTHREAD_CONFIG_MESH_DIAG_ENABLE
+
+template <> otError Interpreter::Process<Cmd("meshdiag")>(Arg aArgs[])
+{
+    otError error = OT_ERROR_NONE;
+
+    /**
+     * @cli meshdiag topology [ip6-addrs][children]
```
Comment (Vyrastas) [inline src/cli/cli.cpp:5427]: Can you remove `[ip6-addrs][children]` here, we'll add them as parameters below

File: src/cli/cli.cpp [hunk 8]
```diff
@@ -5416,6 +5416,214 @@ template <> otError Interpreter::Process<Cmd("nexthop")>(Arg aArgs[])
...
+     *   * `1-links`: Router IDs to which this router has a incoming link with link quality 1
+     *   * If a list if empty, it is omitted in the out.
+     * * If `ip6-addrs`, list of IPv6 addresses of the router
+     * * If `children`, list of all children of the router. Information per child:
+     *   * RLOC16
+     *   * Incoming Link Quality from perspective of parent to child (zero indicates unknown)
+     *   * Child Device mode (`r` rx-on-when-idle, `d` Full Thread Device, `n` Full Network Data, `-` no flags set)
+     *   * Whether the child is this device itself (`me`)
+     *   * Whether the child acts as a border router providing external connectivity (`br`)
+     */
```
Comment (Vyrastas) [inline src/cli/cli.cpp:5471]: Please also add command parameters and a link to the primary API being used

```
* @cparam meshdiag topology [@ca{ip6-addrs}] [@ca{children}]
* @sa otMeshDiagDiscoverTopology
```

File: src/cli/cli.cpp (hunk 4 above)
Comment (Vyrastas) [reply src/cli/cli.cpp:5472]: Yes they will work.  Added a few comments to complete the documentation for the command.

**Review by Vyrastas — CHANGES_REQUESTED** [review]
Let's sort these out and I'm good

File: src/cli/cli.cpp (hunk 8 above)
Comment (abtink) [reply src/cli/cli.cpp:5472]: Done

**Review by Vyrastas — APPROVED** [review]

**Review by jwhui — COMMENTED** [review]
👍🏼

File: src/core/utils/mesh_diag.cpp (hunk 2 above)
Comment (ekobres) [reply src/core/utils/mesh_diag.cpp:182]: @abtink This is a well-tested pattern: retry with exponential backoff. The increasing distance between retries of different devices is a feature of this pattern and not a bug. (https://en.wikipedia.org/wiki/Exponential_backoff)

**Pull request merged**